CRM/Core/DAO.php

```diff
@@ -39,6 +39,13 @@ class CRM_Core_DAO extends DB_DataObject {
    */
   public static $_primaryKey = ['id'];
 
+  /**
+   * @return string[]
+   */
+  protected function getPrimaryKey(): array {
+    return static::$_primaryKey;
+  }
```

Comment:
There seems to be a tension in that:

* SQL tables *can* have composite primary keys
* Civi SQL tables (DAOs) never (almost never?) have composite primary keys. (*At least, I can't think of any off the top of my head...*)

I feel like the question is whether the system permits composite PKs:

* If composite PKs are allowed, then `getPrimaryKey(): string[]` (as above) is good, but the other references to `[0]` invite quiet bugs.
* If composite PKs are not allowed, then `getPrimaryKey(): string` would be better. All the other `[0]` bits can be cleaned out, and there's no risk of them being mistaken. 
    * (Probably `getPrimaryKey()` should embed a validation -- ie complain if there is composite PK.)

Reply:
Yeah - we do permit composite `primaryKeys` - I have an extenstion that does that - but they won't be working here before or after & if feels like a big risk of scope blow out to fix a hypothetical problem (hypothetical that they would ever expect this code path to work for them) if we worry about that - hence I just commented to at least give a clue for someone who might care in the future

Reply:
(the main support we provide for composite primaryKeys at the moment is in GenCode generating the sql)

Reply:
Here's how APIv4 handles it:
- At the lowest level, APIv4 admits the possibility composite keys and `$entity[primary_key]` is passed around as an array.
- But the helper function `CoreUtil::getIdFieldName()` returns a string (the first item in the array), because that's all the API really cares about.

Since this is a low-level function, I guess returning the array is ok. At least it leaves the door open to improve support for composite keys sometime in the future.

I recommend renaming the function `getPrimaryKeys` for clarity.

Reply:
I'm OK with `getPrimaryKeys` - I don't prefer it, because the property is `primaryKey` - but I'm OK to rename it if it is mergeable with that

Reply:
Ok you're right let's stick with matching the property name.

Reply:
I agree with wanting to punt on how to fix composite-PKs. NB I think it's clear that composite-PKs will not behave *well*, and (to a first skim/guess) they probably behaves *badly* (eg  overwriting unrelated records).

The best way to punt on an unsupported scenario is throw an error...

I've tacked a commit on #24287 to make this safer.